thewalrus/quantum/conversions.py

```diff
@@ -0,0 +1,187 @@
+    if isinstance(modes, int):
+        modes = [modes]
+
+    if np.any(np.array(modes) > N):
+        raise ValueError("Provided mode is larger than the number of subsystems.")
+
+    if len(modes) == N:
+        # reduced state is full state
+        return mu, cov
+
+    ind = np.concatenate([np.array(modes), np.array(modes) + N])
+    rows = ind.reshape(-1, 1)
+    cols = ind.reshape(1, -1)
+
+    return mu[ind], cov[rows, cols]
+
+
+def Xmat(N):
+    r"""Returns the matrix :math:`X_n = \begin{bmatrix}0 & I_n\\ I_n & 0\end{bmatrix}`
+
+    Args:
+        N (int): positive integer
+
+    Returns:
+        array: :math:`2N\times 2N` array
+    """
+    I = np.identity(N)
+    O = np.zeros_like(I)
+    X = np.block([[O, I], [I, O]])
+    return X
+
+
+def Qmat(cov, hbar=2):
+    r"""Returns the :math:`Q` Husimi matrix of the Gaussian state.
+
+    Args:
+        cov (array): :math:`2N\times 2N xp-` Wigner covariance matrix
+        hbar (float): the value of :math:`\hbar` in the commutation
+            relation :math:`[\x,\p]=i\hbar`.
+
+    Returns:
+        array: the :math:`Q` matrix.
+    """
+    # number of modes
+    N = len(cov) // 2
+    I = np.identity(N)
+
+    x = cov[:N, :N] * 2 / hbar
+    xp = cov[:N, N:] * 2 / hbar
+    p = cov[N:, N:] * 2 / hbar
+    # the (Hermitian) matrix elements <a_i^\dagger a_j>
+    aidaj = (x + p + 1j * (xp - xp.T) - 2 * I) / 4
+    # the (symmetric) matrix elements <a_i a_j>
+    aiaj = (x - p + 1j * (xp + xp.T)) / 4
+
+    # calculate the covariance matrix sigma_Q appearing in the Q function:
+    # Q(alpha) = exp[-(alpha-beta).sigma_Q^{-1}.(alpha-beta)/2]/|sigma_Q|
+    Q = np.block([[aidaj, aiaj.conj()], [aiaj, aidaj.conj()]]) + np.identity(2 * N)
+    return Q
+
+
+def Covmat(Q, hbar=2):
+    r"""Returns the Wigner covariance matrix in the :math:`xp`-ordering of the Gaussian state.
+    This is the inverse function of Qmat.
+
+    Args:
+        Q (array): :math:`2N\times 2N` Husimi Q matrix
+        hbar (float): the value of :math:`\hbar` in the commutation
+            relation :math:`[\x,\p]=i\hbar`.
+
+    Returns:
+        array: the :math:`xp`-ordered covariance matrix in the xp-ordering.
+    """
+    # number of modes
+    n = len(Q) // 2
+    I = np.identity(n)
+    N = Q[0:n, 0:n] - I
+    M = Q[n : 2 * n, 0:n]
+    mm11a = 2 * (N.real + M.real) + np.identity(n)
+    mm22a = 2 * (N.real - M.real) + np.identity(n)
+    mm12a = 2 * (M.imag + N.imag)
+    cov = np.block([[mm11a, mm12a], [mm12a.T, mm22a]])
+
+    return (hbar / 2) * cov
+
+
+def Amat(cov, hbar=2, cov_is_qmat=False):
+    r"""Returns the :math:`A` matrix of the Gaussian state whose hafnian gives the photon number probabilities.
+
+    Args:
+        cov (array): :math:`2N\times 2N` covariance matrix
+        hbar (float): the value of :math:`\hbar` in the commutation
+            relation :math:`[\x,\p]=i\hbar`.
+        cov_is_qmat (bool): if ``True``, it is assumed that ``cov`` is in fact the Q matrix.
```

Comment:
This is another inconsistency; the `Amat` function accepts either a covariance matrix or Q matrix depending on keyword arguments.

However, we have two separate `Qmat` functions; one that accepts an adjacency matrix, and another accepting a covariance matrix.

Reply:
Are you suggesting to merge the later or separate the former?

Reply:
I don't know 😆 just something I noticed. Would you have a preference?